src/ModelContextProtocol/Shared/McpJsonRpcEndpoint.cs

```diff
@@ -22,356 +16,99 @@ namespace ModelContextProtocol.Shared;
 /// </summary>
 internal abstract class McpJsonRpcEndpoint : IAsyncDisposable
 {
-    private readonly ITransport _transport;
-    private readonly ConcurrentDictionary<RequestId, TaskCompletionSource<IJsonRpcMessage>> _pendingRequests;
-    private readonly ConcurrentDictionary<string, List<Func<JsonRpcNotification, Task>>> _notificationHandlers;
-    private readonly Dictionary<string, Func<JsonRpcRequest, CancellationToken, Task<object?>>> _requestHandlers = [];
-    private int _nextRequestId;
-    private readonly JsonSerializerOptions _jsonOptions;
-    private bool _isDisposed;
+    private readonly RequestHandlers _requestHandlers = [];
+    private readonly NotificationHandlers _notificationHandlers = [];
+
+    private McpSession? _session;
+    private CancellationTokenSource? _sessionCts;
+    private int _started;
+
+    private readonly SemaphoreSlim _disposeLock = new(1);
+    private bool _disposed;
 
     protected readonly ILogger _logger;
 
     /// <summary>
     /// Initializes a new instance of the <see cref="McpJsonRpcEndpoint"/> class.
     /// </summary>
-    /// <param name="transport">An MCP transport implementation.</param>
     /// <param name="loggerFactory">The logger factory.</param>
-    protected McpJsonRpcEndpoint(ITransport transport, ILoggerFactory? loggerFactory = null)
+    protected McpJsonRpcEndpoint(ILoggerFactory? loggerFactory = null)
     {
-        Throw.IfNull(transport);
-
-        _transport = transport;
-        _pendingRequests = new();
-        _notificationHandlers = new();
-        _nextRequestId = 1;
-        _jsonOptions = McpJsonUtilities.DefaultOptions;
         _logger = loggerFactory?.CreateLogger(GetType()) ?? NullLogger.Instance;
     }
 
-    /// <summary>
-    /// Gets whether the endpoint is initialized and ready to process messages.
-    /// </summary>
-    public bool IsInitialized { get; set; }
+    protected void SetRequestHandler<TRequest, TResponse>(string method, Func<TRequest?, CancellationToken, Task<TResponse>> handler)
+        => _requestHandlers.Set(method, handler);
+
+    public void AddNotificationHandler(string method, Func<JsonRpcNotification, Task> handler)
+        => _notificationHandlers.Add(method, handler);
```

Comment:
Do we still need these, or could we refactor things (maybe after this PR) so that all configuration, including handlers, are passed into the ctor? Right now I think we have a concurrency bug, where an incoming notification could trigger enumerating a handler list while it's being modified. It'd be nice to avoid that by design.

Reply:
I think I see where you're going, but I'm not 100% sure. I like the idea of factoring the code in a way where we avoid race conditions by design. It seemed to me that NotificationHandlers unlike RequestHandlers were designed to be more flexible with when you could register them hence the ConcurrentDictionary. I could see that being useful, but only if it can be done soundly. Regardless, I don't think my change makes any NotificationHandler concurrency bugs worse, so I think we could address this in a follow up PR.

Reply:
Basically add to McpServerOptions / McpClientOptions the dictionary of notification handlers at create time rather than adding them one by one after construction / creation. I'm not sure what the scenario is that requires adding handlers after connecting, especially since there's no mechanism provided to remove them. And adding them after connection is inherently racy, such that notifications can be missed if you add the handler too late.

Reply:
(But we can do that after this PR.)